spec.bs

```diff
@@ -6,57 +6,417 @@ Status: CG-DRAFT
+</xmp>
 
-[Exposed=(Window)]
-interface SharedStorage {
-  Promise<any> set(DOMString key, DOMString value, optional SharedStorageSetMethodOptions options = {});
+The {{SharedStorageWorklet}} object allows developers to supply scripts to process Shared Storage data on a separate thread and then output the result through one or more of the output gates. Currently there are two output gates, the private aggregation output gate and the URL selection output gate.
 
-  Promise<any> append(DOMString key, DOMString value);
+The {{SharedStorageWorklet}}'s worklet global scope type is {{SharedStorageWorkletGlobalScope}}. The scripts should define and {{register}} one or more {{SharedStorageOperation}}s. 
 
-  Promise<any> delete(DOMString key);
+<xmp class='idl'>
+  [Exposed=SharedStorageWorklet, Global=SharedStorageWorklet]
+  interface SharedStorageWorkletGlobalScope : WorkletGlobalScope {
+    undefined register(DOMString name,
+                       SharedStorageOperationConstructor operationCtor);
+  };
+
+  callback SharedStorageOperationConstructor = SharedStorageOperation(object options);
+
+  [Exposed=SharedStorageWorklet]
+  interface SharedStorageOperation {
+  };
+</xmp>
+
+Currently each {{SharedStorageOperation}} registered via {{SharedStorageWorkletGlobalScope/register()}} must be one of the following two types:
+* {{SharedStorageRunOperation}}
+* {{SharedStorageSelectURLOperation}}
+
+A {{SharedStorageRunOperation}} is an {{SharedStorageOperation}} that has no JavaScript return value, and hence is especially suited for calling into the private aggregation service.
+
+A {{SharedStorageSelectURLOperation}} is an {{SharedStorageOperation}} that takes in a list of {{SharedStorageUrlWithMetadata}} dictionaries (i.e. URLs each wrapped with metadata) and returns a {{long}} integer index specifying which of these URLs should be selected.
+
+<xmp class='idl'>
+  [Exposed=SharedStorageWorklet]
+  interface SharedStorageRunOperation : SharedStorageOperation {
+    Promise<undefined> run(object data);
+  };
 
-  Promise<any> clear();
+  [Exposed=SharedStorageWorklet]
+  interface SharedStorageSelectURLOperation : SharedStorageOperation {
+    Promise<long> run(object data,
+                      FrozenArray<SharedStorageUrlWithMetadata> urls);
+  };
 
-  Promise<USVString> selectURL(DOMString name,
+  dictionary SharedStorageUrlWithMetadata {
+    USVString url;
+    object reportingMetadata;
+  };
+</xmp>
+
+  ## Registering a Worklet ## {#register-worklet}
+
+  <div algorithm>
+    The <dfn method for="SharedStorageWorkletGlobalScope">register(|name|, |operationCtor|)</dfn> method steps are:
+
+    1. If |name| is missing or empty, throw a {{TypeError}}.
+    1. Let |operation_map| be the internal operation <a href="https://infra.spec.whatwg.org/#ordered-map">map</a> for this scope.
+    1. If |name| already exists as a key in |operation_map|, throw a {{TypeError}}.
+    1. If |operationCtor| is missing or not a constructor, throw a {{TypeError}}.
```

Comment:
@domenic What is the correct way to test if a variable contains a javascript constructor?  Similarly for accessing the prototype, etc, below.  Thanks,

Reply:
Updated. Let me know what you think.